aptos/contracts/sources/governance/contract_upgrade.move

```diff
@@ -0,0 +1,78 @@
+module pyth::contract_upgrade {
+    use wormhole::cursor;
+    use pyth::deserialize;
+    use pyth::contract_upgrade_hash::{Self, Hash};
+    use pyth::state::{Self};
+    use std::vector;
+    use std::aptos_hash;
+    use aptos_framework::code;
+    use pyth::error;
+
+    friend pyth::governance;
+
+    const HASH_LENGTH: u64 = 32;
+
+    struct AuthorizeContractUpgrade {
+        hash: Hash,
+    }
+
+    public(friend) fun execute(payload: vector<u8>) {
+        let AuthorizeContractUpgrade {hash: hash} = from_byte_vec(payload);
+        state::set_contract_upgrade_authorized_hash(hash)
+    }
+
+    fun from_byte_vec(bytes: vector<u8>): AuthorizeContractUpgrade {
+        let cursor = cursor::init(bytes);
+        let hash = contract_upgrade_hash::from_byte_vec(deserialize::deserialize_vector(&mut cursor, HASH_LENGTH));
+        cursor::destroy_empty(cursor);
+        
+        AuthorizeContractUpgrade {
+            hash,
+        }
+    }
+
+    public entry fun do_contract_upgrade(
+        metadata_serialized: vector<u8>,
+        code: vector<vector<u8>>,
+    ) {
+        // Check to see if the hash of the given code and metadata matches the authorized hash.
+        // The aptos framework does no validation of the metadata, so we include it in the hash.
+        assert!(matches_hash(code, metadata_serialized, state::get_contract_upgrade_authorized_hash()), error::invalid_upgrade_hash());
+        // Perform the upgrade
+        let pyth = state::pyth_signer();
+        code::publish_package_txn(&pyth, metadata_serialized, code);
```

Comment:
I am not sure how impactful it is but isn't it better to clear up the state at this point? :think:

Reply:
`get_contract_upgrade_authorized_hash` uses `move_from`, so can only be called once.